crates/services/gas_price_service/src/v1/service.rs

```diff
@@ -0,0 +1,510 @@
+                tracing::debug!("Stopping gas price service");
+                should_continue = false;
+            }
+            l2_block_res = self.l2_block_source.get_l2_block() => {
+                tracing::info!("Received L2 block result: {:?}", l2_block_res);
+                let block = l2_block_res?;
+
+                tracing::debug!("Updating gas price algorithm");
+                self.apply_block_info_to_gas_algorithm(block).await?;
+                should_continue = true;
+            }
+            da_block_costs = self.da_source_channel.recv() => {
+                tracing::info!("Received DA block costs: {:?}", da_block_costs);
+                let da_block_costs = da_block_costs?;
+
+                tracing::debug!("Updating DA block costs");
+                self.apply_da_block_costs_to_gas_algorithm(da_block_costs).await?;
+                should_continue = true;
+            }
+        }
+        Ok(should_continue)
+    }
+
+    async fn shutdown(mut self) -> anyhow::Result<()> {
+        // handle all the remaining l2 blocks
+        while let Some(Ok(block)) = self.l2_block_source.get_l2_block().now_or_never() {
+            tracing::debug!("Updating gas price algorithm");
+            self.apply_block_info_to_gas_algorithm(block).await?;
+        }
+
+        while let Ok(da_block_costs) = self.da_source_channel.try_recv() {
+            tracing::debug!("Updating DA block costs");
+            self.apply_da_block_costs_to_gas_algorithm(da_block_costs)
+                .await?;
+        }
+
+        // run shutdown hooks for internal services
+        self.da_source_adapter_handle.shutdown().await?;
+
+        Ok(())
+    }
+}
+
+pub fn initialize_algorithm<Metadata>(
+    config: &V1AlgorithmConfig,
+    latest_block_height: u32,
+    metadata_storage: &Metadata,
+) -> crate::common::utils::Result<(AlgorithmUpdaterV1, SharedV1Algorithm)>
+where
+    Metadata: MetadataStorage,
+{
+    let algorithm_updater;
+    if let Some(updater_metadata) = metadata_storage
+        .get_metadata(&latest_block_height.into())
+        .map_err(|err| {
+            crate::common::utils::Error::CouldNotInitUpdater(anyhow::anyhow!(err))
+        })?
+    {
+        if let Ok(v1_metadata) = V1Metadata::try_from(updater_metadata.clone()) {
+            algorithm_updater = v1_algorithm_from_metadata(v1_metadata, config);
+        } else {
+            let v0_metadata = V0Metadata::try_from(updater_metadata).map_err(|_| {
+                crate::common::utils::Error::CouldNotInitUpdater(anyhow::anyhow!(
+                    "Could not convert metadata to V0Metadata"
+                ))
+            })?;
+            let v1_metadata = V1Metadata::construct_from_v0_metadata(v0_metadata, config)
+                .map_err(|err| {
+                    crate::common::utils::Error::CouldNotInitUpdater(anyhow::anyhow!(err))
+                })?;
+            algorithm_updater = v1_algorithm_from_metadata(v1_metadata, config);
+        }
```

Comment:
nit: I think we should just unify this branch and add a method `fn get_v1_metadata(metadata: &Metadata, config: &V1AlgorithmConfig)`. In both branches we end up calling `v1_algorithm_from_metadata`, so something like:
```suggestion
        let v1_metadata = get_v1_metadata(updater_metadata, config)?;
        algorithm_updater = v1_algorithm_from_metadata(v1_metadata, config);
```

Reply:
addressed in 2291207dd3